src/pool/srv_util.c

```diff
@@ -316,6 +373,50 @@ ds_pool_map_tgts_update(struct pool_map *map, struct pool_target_id_list *tgts,
 				dom->do_comp.co_status = PO_COMP_ST_DOWNOUT;
 			}
 		}
+		break;
+	}
+
+	return DER_SUCCESS;
+}
+
+/*
+ * Update "tgts" in "map". A new map version is generated only if actual
+ * changes have been made.
+ */
+int
+ds_pool_map_tgts_update(struct pool_map *map, struct pool_target_id_list *tgts,
+			int opc, bool evict_rank)
+{
+	uint32_t	version;
+	int		i;
+	int		rc;
+
+	D_ASSERT(tgts != NULL);
+
+	version = pool_map_get_version(map);
+	for (i = 0; i < tgts->pti_number; i++) {
+		struct pool_target	*target = NULL;
+		struct pool_domain	*dom = NULL;
+
+		rc = pool_map_find_target(map, tgts->pti_ids[i].pti_id,
+					  &target);
+		if (rc <= 0) {
+			D_DEBUG(DF_DSMS, "not find target %u in map %p\n",
+				tgts->pti_ids[i].pti_id, map);
+			continue;
+		}
+
+		dom = pool_map_find_node_by_rank(map, target->ta_comp.co_rank);
+		if (dom == NULL) {
+			D_DEBUG(DF_DSMS, "not find rank %u in map %p\n",
+				target->ta_comp.co_rank, map);
+			continue;
+		}
+
+		rc = update_one_tgt(map, target, dom, opc, evict_rank,
+				    &version);
+		if (rc != 0)
```

Comment:
I wonder should it continue for BUSY here? And also once it failed, should it recover the status. is this suppose to be atomic? @liw

Reply:
This is a great question. 

To add a bit more background info: 
- Currently there are only two cases where it returns BUSY (reintegrate/add a failing node, or drain a reintegrating/adding node). 
- This function only supports one value for opc at a time - so either every change is addition, every change is reintegration, etc. 


If this operation is atomic, then any failure (INVAL or BUSY) will cause no change to the pool map, and the client will get the error. This is the current way this pull request is set up. 

Alternatively, we could allow some operations to proceed. For example, you use DMG to specify 3 ranks should be reintegrated. But one is busy, and two are not. The user could be returned DER_BUSY, but the pool map version will still increment, and reintegration will start for the two non-busy ranks. If the user re-runs their command, the two already-reintegrating ranks will be ignored (whether they just started or have already completed), and the node that was previously busy will start reintegrating. 

The question really is whether the user of DMG would expect a failing command to do nothing, or whether they would appreciate that at least some of their request was started. 

As a user, I probably have a slight preference for atomic transactions because they make the system easier to understand, and easier to script against.

Reply:
oh, ok. if this is always for single target, probably ok for now. But these failure targets needs to be added to the output_list and reply to client?  see ds_pool_update_handler() -->ds_pool_update()->ds_pool_update_internal()?

Reply:
Well, this is multiple targets - but it's for a single operation class (i.e. multiple reintegration targets, multiple rebuild targets, etc. but not a mix). 

Right now the implementation of this is atomic - if you get an error code, all targets you requested failed - so we don't need to return anything back to the client, because the client already knows what they requested and it failed.

If we want to make this non-atomic where some subset of changes succeeds and some fail, then yeah, we probably need a ton of plumbing to be able to return a partial list to the caller. I think this by itself is a pretty compelling reason to leave this as an atomic transaction for now so we don't need to do this extra work.

Reply:
I can't think of any compelling reason forcing us to adopt one way over the other. So I'm fine with the current approach myself.